File: NetFramework/FlexPivot/CS/DataJoin/FormJoin.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Data.SqlClient;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using C1.DataEngine;

namespace DataJoin
{
    public partial class FormJoin : Form
    {
        private List<TreeNode> _tableNodes = new List<TreeNode>();
        public FormJoin()
        {
            InitializeComponent();
            _imageList.Images.Add(Properties.Resources.table_22x16);
            _imageList.Images.Add(Properties.Resources.column_22x16);
            // bold font for table nodes
            tvTables.Font = new System.Drawing.Font(Font, FontStyle.Bold);
        }

        // gets or sets C1DataEngine tables to join
        public dynamic[] Tables { get; set; }

        // gets or sets selected field names to join
        public string[] SelectedFields { get; set; }

        // gets settings object for join
        public Dictionary<string, object> JoinSettings { get; private set; }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);
            ShowTables();
        }

        // saves user choices
        protected override void OnFormClosed(FormClosedEventArgs e)
        {
            SelectedFields = GetCheckedNodes();
            JoinSettings = GetJoinSettings();
        }

        // shows table fields in the tree view
        private void ShowTables()
        {
            TreeNode root = ShowTable(null, Tables[0]);
            ShowTable(root.Nodes[1], Tables[1]);
            ShowTable(root.Nodes[2], Tables[2]);
            CheckNodes();
            tvTables.AfterCheck += tvTables_AfterCheck;
            tvTables.SelectedNode = root;
            tvTables.ExpandAll();
            root.EnsureVisible();
        }

        // shows a table in the tree view
        private TreeNode ShowTable(TreeNode parent, dynamic table)
        {
            TreeNode nt = new TreeNode(table.Table.Name, 0, 0);
            nt.Name = nt.Text;
            nt.Tag = table;
            nt.Checked = true;
            // gets the table as a dictionary of properties
            var props = (IDictionary<string, Object>)table;
            foreach (var prop in props)
            {
                // create a node if the property is a column
                if (prop.Value != null && (prop.Value.GetType() == typeof(Column)))
                {
                    Column col = (Column)prop.Value;
                    TreeNode nc = new TreeNode(col.Name, 1, 1);
                    nc.Name = nc.Text;
                    nc.Tag = col;
                    nc.NodeFont = Font;
                    nt.Nodes.Add(nc);
                }
            }
            // add the table node to the tree view
            // and check referenced nodes
            if (parent != null)
            {
                parent.Nodes.Add(nt);
                parent.Checked = true;
            }
            else
                tvTables.Nodes.Add(nt);

            nt.Nodes[0].Checked = true;
            _tableNodes.Add(nt);
            return nt;
        }

        // creates a settings object to build a join query
        private Dictionary<string, object> GetJoinSettings()
        {
            var settings = new Dictionary<string, object>();
            foreach (TreeNode node in _tableNodes)
            {
                Columns cols = GetSelectedColumns(node);
                if ((object)cols != null)
                    settings[node.Text] = cols;
            }
            return settings;
        }

        // gets selected columns for a table node
        private Columns GetSelectedColumns(TreeNode node)
        {
            Columns cols = null;
            for (int i = 0; i < node.Nodes.Count; i++)
            {
                TreeNode cn = node.Nodes[i];
                // add a selected column to the collection
                if (cn.Checked) 
                {
                    // do not add relationship column to result by default
                    if (node.Parent != null && cn == node.Nodes[0])
                        continue;
                    cols += (Column)cn.Tag;
                }
            }
            if (node.Parent != null)
            {
                // if no result columns are selected, add a relationship column
                if ((object)cols == null)
                    cols += (Column)node.Nodes[0].Tag;
                // add a relation between the tables
                cols |= (Column)node.Parent.Tag == (Column)node.Nodes[0].Tag;
            }
            return cols;
        }

        // gets paths of the selected nodes
        private string[] GetCheckedNodes()
        {
            List<string> nodes = new List<string>();
            GetCheckedNodes(tvTables.Nodes, nodes);
            return nodes.ToArray();
        }

        // adds path of the selected nodes to the list
        private void GetCheckedNodes(TreeNodeCollection nodes, List<string> checkedNodes)
        {
            foreach (TreeNode node in nodes)
            {
                if (node.Checked)
                    checkedNodes.Add(node.FullPath);
                GetCheckedNodes(node.Nodes, checkedNodes);
            }
        }

        // checks tree view nodes by the selected paths
        private void CheckNodes()
        {
            if (SelectedFields == null)
                return;
            foreach (string path in SelectedFields)
                GetNode(path).Checked = true;
        }

        // gets a tree view node by path
        private TreeNode GetNode(string fullPath)
        {
            TreeNode node = null;
            string[] names = fullPath.Split(tvTables.PathSeparator.ToCharArray());
            TreeNodeCollection nodes = tvTables.Nodes;
            foreach (string name in names)
            {
                node = nodes.Find(name, false)[0];
                nodes = node.Nodes;
            }
            return node;
        }

        // validates check state of a node
        // referenced nodes must always be checked
        private void tvTables_AfterCheck(object sender, TreeViewEventArgs e)
        {
            tvTables.AfterCheck -= tvTables_AfterCheck;
            if (e.Node.Nodes.Count > 0)
                e.Node.Checked = true;
            else if (e.Node.Parent != null && e.Node.Parent.Nodes[0] == e.Node)
                e.Node.Checked = true;
            tvTables.AfterCheck += tvTables_AfterCheck;
        }
    }
}

File: NetFramework/FlexPivot/VB/DataJoin/FormJoin.vb
Imports System.Windows.Forms
Imports System.Dynamic
Imports C1.DataEngine

Public Class FormJoin

    Private _tableNodes As New List(Of TreeNode)()
    Public Sub New()
        InitializeComponent()
        _imageList.Images.Add(My.Resources.table_22x16)
        _imageList.Images.Add(My.Resources.column_22x16)
        ' bold font for table nodes
        tvTables.Font = New System.Drawing.Font(Font, FontStyle.Bold)
    End Sub

    ' gets or sets C1DataEngine tables to join
    Public Property Tables() As Object()
        Get
            Return m_Tables
        End Get
        Set(value As Object())
            m_Tables = value
        End Set
    End Property
    Private m_Tables As Object()

    ' gets or sets selected field names to join
    Public Property SelectedFields() As String()
        Get
            Return m_SelectedFields
        End Get
        Set(value As String())
            m_SelectedFields = value
        End Set
    End Property
    Private m_SelectedFields As String()

    ' gets settings object for join
    Public Property JoinSettings() As Dictionary(Of String, Object)
        Get
            Return m_JoinSettings
        End Get
        Private Set(value As Dictionary(Of String, Object))
            m_JoinSettings = value
        End Set
    End Property
    Private m_JoinSettings As Dictionary(Of String, Object)

    Protected Overrides Sub OnLoad(e As EventArgs)
        MyBase.OnLoad(e)
        ShowTables()
    End Sub

    ' saves user choices
    Protected Overrides Sub OnFormClosed(e As FormClosedEventArgs)
        SelectedFields = GetCheckedNodes()
        JoinSettings = GetJoinSettings()
    End Sub

    ' shows table fields in the tree view
    Private Sub ShowTables()
        Dim root As TreeNode = ShowTable(Nothing, Tables(0))
        ShowTable(root.Nodes(1), Tables(1))
        ShowTable(root.Nodes(2), Tables(2))
        CheckNodes()
        AddHandler tvTables.AfterCheck, AddressOf tvTables_AfterCheck
        tvTables.SelectedNode = root
        tvTables.ExpandAll()
        root.EnsureVisible()
    End Sub

    ' shows a table in the tree view
    Private Function ShowTable(parent As TreeNode, table As Object) As TreeNode
        Dim nt As New TreeNode(table.Table.Name, 0, 0)
        nt.Name = nt.Text
        nt.Tag = table
        nt.Checked = True
        ' gets the table as a dictionary of properties
        Dim props = DirectCast(table, IDictionary(Of String, [Object]))
        For Each prop As Object In props
            ' create a node if the property is a column
            If prop.Value IsNot Nothing AndAlso (prop.Value.[GetType]() = GetType(Column)) Then
                Dim col As Column = DirectCast(prop.Value, Column)
                Dim nc As New TreeNode(col.Name, 1, 1)
                nc.Name = nc.Text
                nc.Tag = col
                nc.NodeFont = Font
                nt.Nodes.Add(nc)
            End If
        Next
        ' add the table node to the tree view
        ' and check referenced nodes
        If parent IsNot Nothing Then
            parent.Nodes.Add(nt)
            parent.Checked = True
        Else
            tvTables.Nodes.Add(nt)
        End If

        nt.Nodes(0).Checked = True
        _tableNodes.Add(nt)
        Return nt
    End Function

    ' creates a settings object to build a join query
    Private Function GetJoinSettings() As Dictionary(Of String, Object)
        Dim settings = New Dictionary(Of String, Object)()
        For Each node As TreeNode In _tableNodes
            Dim cols As Columns = GetSelectedColumns(node)
            If DirectCast(cols, Object) IsNot Nothing Then
                settings(node.Text) = cols
            End If
        Next
        Return settings
    End Function

    ' gets selected columns for a table node
    Private Function GetSelectedColumns(node As TreeNode) As Columns
        Dim cols As Columns = Nothing
        For i As Integer = 0 To node.Nodes.Count - 1
            Dim cn As TreeNode = node.Nodes(i)
            ' add a selected column to the collection
            If cn.Checked Then
                ' do not add relationship column by default
                If (node.Parent IsNot Nothing And cn Is node.Nodes(0)) Then
                    Continue For
                End If
                cols += DirectCast(cn.Tag, Column)
            End If
        Next
        If node.Parent IsNot Nothing Then
            ' if no result columns are selected, add a relationship column
            If DirectCast(cols, Object) Is Nothing Then
                cols += DirectCast(node.Nodes(0).Tag, Column)
            End If
            ' add a relation between the tables
            cols = cols Or DirectCast(node.Parent.Tag, Column) = DirectCast(node.Nodes(0).Tag, Column)
        End If
        Return cols
    End Function

    ' gets paths of the selected nodes
    Private Function GetCheckedNodes() As String()
        Dim nodes As New List(Of String)()
        GetCheckedNodes(tvTables.Nodes, nodes)
        Return nodes.ToArray()
    End Function

    ' adds path of the selected nodes to the list
    Private Sub GetCheckedNodes(nodes As TreeNodeCollection, checkedNodes As List(Of String))
        For Each node As TreeNode In nodes
            If node.Checked Then
                checkedNodes.Add(node.FullPath)
            End If
            GetCheckedNodes(node.Nodes, checkedNodes)
        Next
    End Sub

    ' checks tree view nodes by the selected paths
    Private Sub CheckNodes()
        If SelectedFields Is Nothing Then
            Return
        End If
        For Each path As String In SelectedFields
            GetNode(path).Checked = True
        Next
    End Sub

    ' gets a tree view node by path
    Private Function GetNode(fullPath As String) As TreeNode
        Dim node As TreeNode = Nothing
        Dim names As String() = fullPath.Split(tvTables.PathSeparator.ToCharArray(), StringSplitOptions.RemoveEmptyEntries)
        Dim nodes As TreeNodeCollection = tvTables.Nodes
        For Each name As String In names
            node = nodes.Find(name, False)(0)
            nodes = node.Nodes
        Next
        Return node
    End Function

    ' validates check state of a node
    ' referenced nodes must always be checked
    Private Sub tvTables_AfterCheck(sender As Object, e As TreeViewEventArgs)
        RemoveHandler tvTables.AfterCheck, AddressOf tvTables_AfterCheck
        If e.Node.Nodes.Count > 0 Then
            e.Node.Checked = True
        ElseIf e.Node.Parent IsNot Nothing AndAlso e.Node.Parent.Nodes(0) Is e.Node Then
            e.Node.Checked = True
        End If
        AddHandler tvTables.AfterCheck, AddressOf tvTables_AfterCheck
    End Sub

End Class